### PR TITLE
Components: enhancements to TypeScript migration guidelines

### PR DESCRIPTION
**Pull request by ciampo**
## What?

- Add link to TypeScript migration guide in TOC
- Add a point about props destructuring
- Expand point about `WordPressComponentProps`

## Why?

Iteratively improve the TypeScript migration guidelines as we gather insights from the initial work on the matter

### CODE REVIEW

File: packages/components/CONTRIBUTING.md [hunk 1]
```diff
@@ -489,9 +490,23 @@ Given a component folder (e.g. `packages/components/src/unit-control`):
...
+
+			// After:
+			function MyComponent( props ) {
+				const {  myProp1, myProp2, ...restProps } = props;
+
+				/* ... */
+			}
+			```
+
+		3. Add the folders to the `tsconfig.json` file.
```
Comment (ciampo) [inline packages/components/CONTRIBUTING.md:507]: Points 3, 4 and 5 here are unchanged apart from their index, that shiftet by 1 after adding a new point (now no. 2)

File: packages/components/CONTRIBUTING.md [hunk 2]
```diff
@@ -489,9 +490,23 @@ Given a component folder (e.g. `packages/components/src/unit-control`):
...
-		4. Continue with the refactor of the current component (and take care of the refactor of the dependent components at a later stage).
+		2. If the components in the ignored files are destructuring props directly from the function's arguments, move the props destructuring to the function's body (this is to avoid TypeScript errors from trying to infer the props type):
+
+			```jsx
+			// Before:
+			function MyComponent( { myProp1, myProp2, ...restProps } ) { /* ... */ }
+
+			// After:
+			function MyComponent( props ) {
+				const {  myProp1, myProp2, ...restProps } = props;
```
Comment (mirka) [inline packages/components/CONTRIBUTING.md:501]: Good clarification with the example 👍

File: packages/components/CONTRIBUTING.md [hunk 3]
```diff
@@ -509,7 +524,23 @@ Given a component folder (e.g. `packages/components/src/unit-control`):
...
 	6. Extend existing components’ props if possible, especially when a component internally forwards its props to another component in the package.
-	7. Use `WordPressComponent` type if possible.
+	7. If the component forwards its `...restProps` to an underlying element, you should use the `WordPressComponentProps` type for the component's props:
+
+		```jsx
+		import type { WordPressComponentProps } from '../ui/context';
+		import type { ComponentOwnProps } from './types';
+
+		function UnconnectedMyComponent(
+			// The resulting type will include:
```
Comment (mirka) [inline packages/components/CONTRIBUTING.md:524]: Should we add the stuff about the `ref`?

Comment (ciampo) [reply packages/components/CONTRIBUTING.md:524]: Pushed some changes in https://github.com/WordPress/gutenberg/pull/41669/commits/cb34385e89d3ff7885c2ef9fb30570a1f0cb8bb5, let me know what you think!

Main changes:

- moved down the point re. making sure that there's a name export
- added a new point re. ref forwarding (either via `forwardRef` or via `contextConnect`)
- updated index of remaining points as needed

**Comment by walbo** [comment]
Maybe update the section with `@example` as well? The code example doesn't appear in the stories Docs tab if you add `@example` before the code. See https://github.com/WordPress/gutenberg/pull/40915#discussion_r869929288

File: packages/components/CONTRIBUTING.md [hunk 4]
```diff
@@ -489,13 +490,99 @@ Given a component folder (e.g. `packages/components/src/unit-control`):
...
+			// - all props defined in `ComponentOwnProps`
+			// - all HTML props/attributes from the component specified as the second
+			//   parameter (`div` in this example)
+			// - the special `as` prop (which marks the component as polymorphic),
+			//   unless the third parameter is `false`
+			props:  WordPressComponentProps< ComponentOwnProps, 'div', true >
+		) { /* ... */ }
+		```
+
+	7. If the component doesn't forwards its ref yet, wrap the component in a `forwardRed` call. Alternatively, if you want to take advantage of the [Context system](#context-system), you can use the `contextConnect` utility function (which also takes care of adding ref forwarding)
```
Comment (mirka) [inline packages/components/CONTRIBUTING.md:534]:
```suggestion
	7. If the component doesn't forward its ref yet, wrap the component in a `forwardRef` call. Alternatively, if you want to take advantage of the [Context system](#context-system), you can use the `contextConnect` utility function (which also takes care of adding ref forwarding)
```

**Review by mirka — APPROVED** [review]
🚀

**Comment by mirka** [comment]
> Maybe update the section with `@example` as well? The code example doesn't appear in the stories Docs tab if you add `@example` before the code. See [#40915 (comment)](https://github.com/WordPress/gutenberg/pull/40915#discussion_r869929288)

Good point. I was kind of on the fence about showing the `@example` snippets in the Storybook — on one hand they're a bit redundant given the actual story code snippets, but on the other hand the `@example` snippets are more complete with the `import` and `useState` statements. I don't feel strongly either way though. Do y'all have a preference?

File: packages/components/CONTRIBUTING.md (hunk 4 above)
Comment (ciampo) [reply packages/components/CONTRIBUTING.md:534]: Thank you for spotting these typos!

**Comment by ciampo** [comment]
I've updated the point about the `@example` keyword in https://github.com/WordPress/gutenberg/pull/41669/commits/ff88bb5d3438abafdf879feb8d89e22a1d312726 , discouraging its usage.

> Do y'all have a preference?

Let's go with showing the extra snippet in the StoryBook docs, for now. As you say, these snippets can offer a more complete example (with `import`s and `useState`s).

Also, sometimes the dynamic code snippet from StoryBook are not as helpful (e.g. sometimes they show `function noRefCheck(){}` for callback props), so I believe it's probably a good thing to have an extra code snippet.

**Review by walbo — APPROVED** [review]

**Pull request merged**